contributingGuides/review/RULES.md

```diff
@@ -0,0 +1,203 @@
+# ✅ AI Code Review Rules
+
+These rules are used to conduct structured code reviews on pull request diffs. Each rule includes:
+
+- A unique **Rule ID**
+- **Severity level**: MUST / SHOULD / OPTIONAL
+- **Pass/Fail condition**
+- Examples of good and bad usage
+
+---
+
+## Performance Rules
+
+### [PERF-1] No spread in list item's renderItem
+- **Severity**: MUST
+- **Condition**: When passing data to components in renderItem functions, avoid using spread operators to extend objects. Instead, pass the base object and additional properties as separate props to prevent unnecessary object creation on each render.
+
+Good:
+```tsx
+<Component
+  item={item}
+  isSelected={isSelected}
+  shouldAnimateInHighlight={isItemHighlighted}
+/>
+```
+
+Bad:
+```tsx
+<Component
+  item={{
+      shouldAnimateInHighlight: isItemHighlighted,
+      isSelected: selected,
+      ...item,
+  }}
+/>
+```
+
+---
+
+### [PERF-2] Use Set for O(1) lookups instead of array methods
+- **Severity**: SHOULD
+- **Condition**: When performing repeated lookups or checking existence in collections, convert arrays to Sets for O(1) performance instead of using O(n) methods like `.includes()`.
+
+Good:
+```tsx
+const reportIdFromRoute = report.reportId;
+const reportUserIdSet = new Set(Object.values(reports));
+const reportExists = reportUserIdSet.has(reportIdFromRoute);
+```
+
+Bad:
+```tsx
+const reportIdFromRoute = report.reportId;
+const reportExists = Object.values(reports).includes(reportIdFromRoute);
+```
+
+---
+
+### [PERF-3] Use early returns in array iteration methods
+- **Severity**: SHOULD
+- **Condition**: When using `.every()`, `.some()`, or similar methods, perform simple checks first with early returns before expensive operations.
+
+Good:
+```tsx
+const areAllTransactionsValid = transactions.every((transaction) => {
+    if (!transaction.rawData || transaction.amount <= 0) return false;
+    const validation = validateTransaction(transaction);
+    return validation.isValid;
+});
+```
+
+Bad:
+```tsx
+const areAllTransactionsValid = transactions.every((transaction) => {
+    const validation = validateTransaction(transaction);
+    return validation.isValid;
+});
+```
+
+---
+
+### [PERF-4] Use OnyxListItemProvider hooks instead of useOnyx in renderItem
+- **Severity**: MUST
+- **Condition**: Components rendered inside `renderItem` functions should use dedicated hooks from `OnyxListItemProvider` instead of individual `useOnyx` calls.
+
+Good:
+```tsx
+// ReportActionItem.tsx
+const personalDetails = usePersonalDetails();
+```
+
+Bad:
+```tsx
+// ReportActionItem.tsx
+const [personalDetails] = useOnyx(ONYXKEYS.PERSONAL_DETAILS_LIST);
+```
+
+---
+
+### [PERF-5] Memoize objects and functions passed as props
```

Comment:
I don't understand this rule at all, why is useMemo better in this case? Is it because you are creating a smaller object with less props than the full `report`? If so that's not related to memoize really, plus computing the smaller object takes no time, so not sure why we would care about memoizing it.

Reply:
Not really!
The 'good' example's object has really good _referential stability_ (look at its dependencies) meaning it does not change often. 
Bad example will re-render anytime the `lastUpdatedAt` prop on the `report` is going to change, and this can result in some really poor performance.

In fact, the Best solution would just pass each of these primitives as separate props (reportID, type, isPinned). Maybe we should include that as well?

Reply:
Both Good and the mentioned theoretical Best only produce renders when any of these dependency props change, and this is expected to not happen often.

Reply:
ok got it. That will be clearer when we add:
> I think we should add a reasoning or explanation explaining why each rule exists

And explain this 😅